File: src/Python/Problem47.py
"""
The first two consecutive numbers to have two distinct prime factors are:

14 = 2 × 7
15 = 3 × 5

The first three consecutive numbers to have three distinct prime factors are:

644 = 2² × 7 × 23
645 = 3 × 5 × 43
646 = 2 × 17 × 19.

Find the first four consecutive integers to have four distinct prime factors each. What is the first of these numbers?
"""
import math
from Problem35 import sieveOfErat

def listDivisors(n: int) -> list:
    """returns the list of distinct divisors of n"""
    ret = []
    for i in range(1,int(math.sqrt(n)+1)):
        if n%i==0:
            if i not in ret:
                ret.append(i)
            if n/i != i and (n/i not in ret):
                ret.append(int(n/i))
    
    return sorted(ret)

def listPrimeDivisors(n: int) -> list:
    """returns the list of distinct prime divisors of n"""
    alldivs = listDivisors(n) #get all divisors
    maxnum = max(alldivs)
    listprimes = sieveOfErat(maxnum) #get the true list of primes until the highest #
    
    ret = []
    primesindex = 0
    divsindex = 0

    while (primesindex < len(listprimes)) and (divsindex < len(alldivs)):
        if listprimes[primesindex] < alldivs[divsindex]: #move primes pointer up
            primesindex += 1
        elif listprimes[primesindex] > alldivs[divsindex]: #move divs pointer up
            divsindex += 1
        else: #the elements are equal, move both elements up
            ret.append(alldivs[divsindex])
            divsindex += 1
            primesindex += 1

    return ret

def main():
    #print(listDivisors(644))
    #print(listPrimeDivisors(644))
    
    arr = [1,2,3,4]
    while(True):
        #print(arr)
        boolarr = [len(listPrimeDivisors(i))==4 for i in arr]
        if all(boolarr): #all 4 are consecutives that have 4 prime divisors
            print("ans: ",arr[0])
            break

        arr = [i+1 for i in arr]

if __name__ == '__main__':
    main()

File: src/Java/Problem47.java
package Java;
import java.util.*;

public class Problem47
{
    public static void main(String[] args)
    {
        //methods work well
        //System.out.println(listDivisors(644));
        //System.out.println(listPrimeDivisors(644));

        int[] arr = {20,21,22,23};
        while(true)
        {
            boolean flag = false;
            for (int i = 0; i < 4; i++)
            {
                //check each one individually and if any are false, break: no need to do further calculations
                boolean curr = listPrimeDivisors(arr[i]).size() == 4;
                if (!curr)
                {
                    flag = true;
                    break;
                }
            }

            //iterate on to the next
            if (flag)
            {
                for (int i = 0; i < 4; i++)
                    arr[i] += 1;

                continue;
            }

            //if it got here, it never had to break so all 4 entries are good
            System.out.println("answer: " + arr[0]);
            break;
        }
    }

    public static List<Integer> listDivisors(int n)
    {
        List<Integer> ret = new ArrayList<Integer>();
        for (int i = 1; i < (int)(Math.sqrt(n)+1); i++)
        {
            if (n%i==0) //if its a factor
            {
                if (!ret.contains(i)) //and not already in the ret list
                    ret.add(i);
                if (n/i != i && (!ret.contains(n/i))) //same but for the other factor
                    ret.add(n/i);
            }
        }

        Collections.sort(ret); //sort to aid in the 2 pointer method in listPrimeDivisors
        return ret;
    }

    public static List<Integer> listPrimeDivisors(int n)
    {
        List<Integer> alldivs = listDivisors(n);
        int maxnum = alldivs.get(alldivs.size()-1);
        List<Integer> listprimes = Problem35.sieveOfErat(maxnum); //efficient way for large #s

        List<Integer> ret = new ArrayList<Integer>();
        //Use 2 pointer method to do in O(N+M) instead of O(NM), relies off the fact that both are sorted arrays
        int primesindex = 0;
        int divsindex = 0;

        int primessize = listprimes.size();
        int divssize = alldivs.size();

        while(primesindex < primessize && divsindex < divssize)
        {
            if (listprimes.get(primesindex) < alldivs.get(divsindex)) //move prime pointer up
                primesindex++;
            else if (listprimes.get(primesindex) > alldivs.get(divsindex)) //move divs pointer up
                divsindex++;
            else //they are equal, add to the returned list
            {
                ret.add(alldivs.get(divsindex));
                divsindex++;
                primesindex++;
            }
        }

        return ret;
    }
}
